mcstatus/status_response.py

```diff
@@ -0,0 +1,670 @@
+            online=raw["online"],
+            max=raw["max"],
+            sample=sample,
+        )
+
+
+@dataclass
+class BedrockStatusPlayers(BaseStatusPlayers):
+    """Class which represents `BedrockStatusResponse` players field."""
+
+
+@dataclass
+class JavaStatusPlayer:
+    """Class for storing player info. Only for Java."""
+
+    name: str
+    id: str
+
+    @property
+    def uuid(self) -> str:
+        """Alias to the `id` field.
+
+        :return: UUID of the player.
+        """
+        return self.id
+
+    @classmethod
+    def build(cls, raw: Dict[str, Any]) -> Self:
+        """Build `JavaStatusPlayer` from raw response dict.
+
+        :param raw: Raw response dict.
+        :raise ValueError: If the required keys (name, id) are not present.
+        :raise TypeError: If the required keys (name - str, id - str) are not of the specified type.
+        :return: `JavaStatusPlayer` object.
+        """
+        _validate_data(raw, "player", [("name", str), ("id", str)])
+        return cls(name=raw["name"], id=raw["id"])
+
+
+@dataclass
+class BaseStatusVersion(ABC):
+    """Class for storing version info.
+
+    :param name: Version name. Example `1.18.0`.
+    :param protocol: Version protocol. See https://minecraft.fandom.com/wiki/Protocol_version.
+    """
+
+    name: str
+    protocol: int
+
+
+@dataclass
+class JavaStatusVersion(BaseStatusVersion):
+    """Class for storing Java version info."""
+
+    @classmethod
+    def build(cls, raw: Dict[str, Any]) -> Self:
+        """Build `JavaStatusVersion` from raw response dict.
+
+        :param raw: Raw response dict.
+        :raise ValueError: If the required keys (name, protocol) are not present.
+        :raise TypeError: If the required keys (name - str, protocol - int) are not of the specified type.
+        :return: `JavaStatusVersion` object.
+        """
+        _validate_data(raw, "version", [("name", str), ("protocol", int)])
+        return cls(name=raw["name"], protocol=raw["protocol"])
+
+
+@dataclass
+class BedrockStatusVersion(BaseStatusVersion):
+    """Class for storing Bedrock version info.
+
+    :param brand: Like `MCPE` or another.
+    """
+
+    brand: str
+
+
+def custom_eq(self: Any, other: Any) -> bool:  # noqa: ANN401 # actually will work with any dataclass
+    """Custom equality function for outdated subclasses."""
+    to_check = list(
+        (
+            (getattr(self, self_field.name) == getattr(other, other_field.name))
+            if self_field.compare and other_field.compare and (self_field.name == other_field.name)
+            else True
+        )
+        for self_field in fields(self)
+        for other_field in fields(other)
+    )
+    to_check.append(isinstance(self, other.__class__))
+    return all(to_check)
+
+
+_OLD_JAVA_INIT_SIGNATURE = Signature(
+    parameters=[
+        Parameter("raw", Parameter.POSITIONAL_OR_KEYWORD, annotation=Dict[str, Any]),
+    ]
+)
+
+
+class JavaStatusResponse(__JavaStatusResponse):
```

Comment:
We already have [disscussion on this theme](https://github.com/py-mine/mcstatus/pull/306#discussion_r889579462), let's keep one disscusion for one theme.

I would like to only add: please read [this message](https://github.com/py-mine/mcstatus/pull/306#discussion_r872993704), I already answered there on all questions, that you ask about my deprecations system.

Reply:
> Also, just keeping original names is impossible, I didn't just move classes, and rename some fields. ...

I'm not saying to keep original names of the classes, just of the variables they held. 

> Users can initialize these classes in tests ...

As for the usecase of initializations in tests, yes, there indeed are some people who did do that, however kevin has already mentioned this, and found that it wasn't very common (using a github search).

> I don't see any way how to do this PR without breaking changes, or at least deprecations.

You can use some deprecations on variables where it's needed, I'm just saying deprecations of `__init__` are not all that necessary, as we don't consider them part of public API. Most of the deprecations done here are precisely to address these signatures, and I'm pretty sure this is what kevin was talking about.

Reply:
If only `__init__` method... But I already did the deprecations, so should we spend time on trifle things, or review the PR, so it will be finally merged?

Reply:
> But I already did the deprecations

I'm aware that the deprecations are already implemented now, but that can still be reverted. I think this is the way we should handle it, so that this PR actually meets it's title, saying it's a refactor. If you still want to see the renames introduced, you're free to open another PR for it, or perhaps an issue first, so that we can discuss whether we even want this rename, as it seems like a pretty major point of conflict here and probably the primary reason this PR is still not yet merged.

Having changes like renames or reimplementations that require deprecations in a standalone PR is a much better option anyway, as it allows us to reference it easily to people who ask about why it was done, without that PR also containing a ton of other changes, making these changes harder to find in it.

> so should we spend time on trifle things, or review the PR, so it will be finally merged?

Remember that this project is maintained by volunteers that do it for free in their spare time, please respect and understand that we often can't invest that much time into this project, especially for these bigger PRs that take a lot of time to review. 

Also, the code quality is pretty important for us, and so it's not surprising that a request to heavily change this internal structure will take a while. This PR isn't a critical bugfix or security issue to the project, and it doesn't need to get merged particularly quickly, instead we should focus on doing things right, so that anyone looking over the codebase will have an easy time understanding things.

Reply:
> I'm aware that the deprecations are already implemented now, but that can still be reverted.

For what? I would agree, if we will discuss it for implementing less code because this part is partially not a part of public API, but still, these functions can be easily used for building objects in tests or other situations which we discussed somewhere else.

**BUT**, we are discussing this *after* I did the implementation, so I don't see sense in removing additional deprecations of partially public API, which still can be used by users.

> I think this is the way we should handle it, so that this PR actually meets it's title, saying it's a refactor.

I [already said that fully following definition](https://github.com/py-mine/mcstatus/pull/306#discussion_r947897925) provided by @kevinkjt2000:

> > Refactor - To [rewrite](https://www.wordnik.com/words/rewrite) existing [source code](https://www.wordnik.com/words/source%20code) in order to [improve](https://www.wordnik.com/words/improve) its [readability](https://www.wordnik.com/words/readability), [reusability](https://www.wordnik.com/words/reusability) or [structure](https://www.wordnik.com/words/structure) without [affecting](https://www.wordnik.com/words/affecting) its [meaning](https://www.wordnik.com/words/meaning) or [behaviour](https://www.wordnik.com/words/behaviour)
>
> I'm fully following this definition:
> - To [rewrite](https://www.wordnik.com/words/rewrite) existing [source code](https://www.wordnik.com/words/source%20code): old classes were fully removed and moved to its own file.
> - in order to [improve](https://www.wordnik.com/words/improve) its [readability](https://www.wordnik.com/words/readability), [reusability](https://www.wordnik.com/words/reusability) or [structure](https://www.wordnik.com/words/structure): inheritance is more readable, reusable, and structured than duplicated code.
> - without [affecting](https://www.wordnik.com/words/affecting) its [meaning](https://www.wordnik.com/words/meaning) or [behavior](https://www.wordnik.com/words/behaviour): goal of these classes the same as was. I also didn't touch the logic - constructors was rewritten, but works the same, as before.

> If you still want to see the renames introduced, you're free to open another PR for it, or perhaps an issue first, so that we can discuss whether we even want this rename

What is the sense to return old and different names? This situation fully looks like:

```python
# was
class FirstAnswer:
    online_num: int
    max_online: int
    status_code: int
    description: str

class OnlineFieldInSecondAnswer:
    current: int
    max: int

class SecondAnswer:
    online: OnlineFieldInSecondAnswer
    status_code_number: int
    custom_info: list

# now
class OnlineInAnswer:
    current: int
    max: int

class AbstractAnswer(ABC)
    online: OnlineInAnswer
    status_code: int

class FirstAnswer(AbstractAnswer):
    description: str

class SecondAnswer(AbstractAnswer):
    custom_info: list
```

See? I renamed fields to **one** and **common** names. Do you really want to see inside this code something like:

```python
class FirstAnswer(AbstractAnswer):
    description: str

    @property
    def online_num(self):
        return self.online.current

    @property
    def max_online(self):
        return self.online.max

class SecondAnswer(AbstractAnswer):
    custom_info: list

    @property
    def status_code_number(self):
        return self.status_code
```

Now do you see how is dirty code after returning old names, just as useless aliases? Also note that this example only uses 2 classes and 4 fields in each. I have much larger numbers, so the code size will increase at a very high rate.

> as it seems like a pretty major point of conflict here and probably the primary reason this PR is still not yet merged.

I don't agree. Personally, I think that this PR is still not yet merged because of many changes, that show GitHub. I will return to this theme later.

> Having changes like renames or reimplementations that require deprecations in a standalone PR is a much better option anyway

I agree, but I can't open PR in my own branch. If these changes will be directed to branch in `py-mine/mcstatus`, I would appreciate the idea about slicing this PR on multiple parts. However, this is too late, for moving this PR because it already has big history of discussions and commits.

> Remember that this project is maintained by volunteers that do it for free in their spare time, please respect and understand that we often can't invest that much time into this project

I'm sorry if my comment somehow looks offending. I didn't mean anything like this, maybe something wrong with `trifle` word (because I actually took it from translator). Of course, I respect your contributions and reviews in this PR, but it's been a long time since I opened the PR.

I meant there is a topic, described in the first paragraph of this comment. Like, you should spend your time on reviewing more important things, instead of such useless theme (as I described in first paragraph of this comment).

> especially for these bigger PRs that take a lot of time to review.

This PR is not big. Really, look at `status_response.py`. Only on 95th string there are some real code, all before - just types and imports. Then after 352nd up to end of the file (734th line) there are going deprecations logic, which contain a lot of duplicated code (I can't move it to its own function). You can also look to `pinger.py` file, there I moved `TypedDict`s for `description` field plus removed old `PingResponse`, and in result **161 lines**! Looks like a count for one PR, but I just moved those in new file!

As you see, numbers of changed lines inside diff, don't say anything about how big is a PR. Even one line change can break everything, and require tons of changes in user code.

> the code quality is pretty important for us

It's super. This is awesome, and I see it in how do you review PRs.

> it's not surprising that a request to heavily change this internal structure will take a while.

However, I didn't change anything important. I only changed interface, and moved a lot of things, so it's now on the right place, with new structure.

> instead we should focus on doing things right, so that anyone looking over the codebase will have an easy time understanding things.

But the theme for this conversation isn't something, that you should to focus now.

Reply:
> BUT, we are discussing this after I did the implementation, so I don't see sense in removing additional deprecations of partially public API, which still can be used by users.

The simple reason is that the resulting code will be less complex and hence much easier to read. Using this kind of deprecation strategy is very uncommon and while it does work well, it can be a bit hard to orient in, especially for potential new contributors, and since our deprecation periods tend to be relatively long, it would mean the messy logic would stay in for a long time. Considering these custom deprecations are pretty much pointless as these objects aren't a part of the public API, there is just no real reason to keep them there.

> Now do you see how is dirty code after returning old names, just as useless aliases? Also note that this example only uses 2 classes and 4 fields in each. I have much larger numbers, so the code size will increase at a very high rate.

I don't find these that bad, properties are pythonic and there's nothing wrong with those, yes they may be a bit more annoying here considering they're just aliases, but again, that might be a thing we can change in another PR. Let's not do everything in one place.

> I agree, but I can't open PR in my own branch. If these changes will be directed to branch in py-mine/mcstatus, I would appreciate the idea about slicing this PR on multiple parts. However, this is too late, for moving this PR because it already has big history of discussions and commits.

That's not necessarily the case. You can still remove the renames from this PR, wait until it gets merged and open another PR to `master` afterwards. 

Yes, it does mean we won't be able to review it at the same time as this one as it depends on this getting merged, but it's not a critical change and it can wait. As this PR will become much easier to review and it will mean we'll be able to immediately close some still unresolved conversations about these renames, I'd expect that this wouldn't take as long, especially since a lot of reviewing has already been done here, so there isn't that much left.

> As you see, numbers of changed lines inside diff, don't say anything about how big is a PR. Even one line change can break everything, and require tons of changes in user code.

Logic wise, that's true, the number of lines isn't too important, but review-wise, we need to check every line and ensure it meets our code quality requirements and that it makes sense, not to mention reviewing the complex deprecation strategy here. The PR also adds/changes a lot of tests which we need to ensure are well written and cover all edge cases, etc. So there really is a lot to review, even though the change in logic is relatively simple to explain.

Reply:
I understood your idea more after explanation in Discord. I would agree to slice this PR on three parts (or two if move last one):
1. Main logic.
2. Deprecations.
3. Tests (however, shouldn't it go with the first point?).

Except for the decision when we remove all deprecations, and just bumping major version once more. But this should be discussed more here, or in Discord [(see also my last unanswered answer)](https://discord.com/channels/936788458939224094/938623566964981802/1025799067743309844).

And I am completely against merging those "slices" to `master`. Much better will be just create one branch for those slices and #335 (fortunately, GitHub supports change of the target branch).

Reply:
Deprecations were removed in f3dfdf87712c5d8676989e5aae72a9cf4c3a7854. Could it be closed?

(I think my last message can be easily discussed in Discord.)